tests/test_software_upgrade.sh

```diff
@@ -0,0 +1,169 @@
+
+    # Set time to wait for proposal to pass
+    voting_period=$(curl -s http://localhost:26657/genesis\? | jq -r '.result.genesis.app_state.gov.voting_params.voting_period')
+    # voting_period=${voting_period::-1}
+    echo "Adding 5s to the voting period ($voting_period) to calculate the upgrade height."
+    voting_period_seconds=${voting_period::-1}
+    let voting_waiting_time=$voting_period_seconds+5
+
+    # Calculate upgrade height
+    let voting_blocks_delta=$voting_period_seconds/5+3
+    height=$(curl -s http://$gaia_host:$gaia_port/block | jq -r .result.block.header.height)
+    upgrade_height=$(($height+$voting_blocks_delta))
+    echo "Upgrade block height set to $upgrade_height."
+
+    # Get validator address and account sequence
+    # Sequence number must be added 1 for the vote to be accepted.
+    echo "Querying gaiad for the validator address and account sequence..."
+    validator_address=$(gaiad keys list --keyring-backend test --output json | jq -r '.[0].address')
+    validator_sequence=$[ $(gaiad query auth account $validator_address --output json | jq -r '.sequence')+1 ]
+    echo "The validator has address $validator_address, setting sequence # to $validator_sequence."
+
+    # Manual download
+    # echo "Downloading gaiad binary for $upgrade_version..."
+    # upgrade_info = "-"
+    # mkdir -p ~/.gaia/cosmovisor/upgrades/$upgrade_name/bin
+    # wget https://github.com/cosmos/gaia/releases/download/$upgrade_version/gaiad-$upgrade_version-linux-amd64 -O ~/.gaia/cosmovisor/upgrades/$upgrade_name/bin/gaiad
+    # chmod +x ~/.gaia/cosmovisor/upgrades/$upgrade_name/bin/gaiad
+    # proposal="gaiad tx gov submit-proposal software-upgrade $upgrade_name --from $validator_address --keyring-backend test --upgrade-height $upgrade_height --title gaia-upgrade --description 'test' --chain-id my-testnet --deposit 10stake --yes"
+
+    # Auto download: Set the binary paths need for the proposal message
+    download_path="https://github.com/cosmos/gaia/releases/download/$upgrade_version"
+    upgrade_info="{\"binaries\":{\"linux/amd64\":\"$download_path/gaiad-$upgrade_version-linux-amd64\",\"linux/arm64\":\"$download_path/$upgrade_version/gaiad-$upgrade_version-linux-arm64\",\"darwin/amd64\":\"$download_path/gaiad-$upgrade_version-darwin-amd64\",\"windows/amd64\":\"$download_path/gaiad-$upgrade_version-windows-amd64.exe\"}}"
+    proposal="gaiad tx gov submit-proposal software-upgrade $upgrade_name --from $validator_address --keyring-backend test --upgrade-height $upgrade_height --upgrade-info $upgrade_info --title gaia-upgrade --description 'test' --chain-id my-testnet --deposit 10stake --yes"
+
+    # Submit the proposal and vote yes for it
+    echo "Submitting the upgrade proposal."
+    echo $proposal
+    $proposal
+
+    echo "Submitting the \"yes\" vote."
+    vote="gaiad tx gov vote 1 yes --from $validator_address --keyring-backend test --chain-id my-testnet --sequence $validator_sequence --yes"
```

Comment:
Curious why "sequence" is required?

Reply:
This is related to the previous comment: if I submit the proposal without the `--sequence` flag I get an error saying the sequence number did not match the expected one. But if I set that sequence value to the one given by the `gaiad q auth account` command, it is still always off, by one. That's why I ended up adding 1 to it in the assignment line.
I don't know if it's a bug, but I'll test it a couple more times and bring it up with the IG team.